### DISCUSSION TITLE
Stepper Height

### DISCUSSION THREAD

**Discussion by natankeddem** [zauberzeug/nicegui#1788] [Q&A]
### Question

I have been banging away at a specific layout I can't seem to make work. This is a more minimal example of the problem I am facing, in the actual code there is much more inside including a scroll with multiple other controls. In most cases before this I was able to get full height for containers/elements using the `col` class. Basically I have a stepper with 2 inner elements which are `justify-between` so `input` on top and `button` on bottom. The stepper has been tricky to expand down unless the inner elements have a fixed height. I have this working with a `calc` and some magic numbers but was hoping for something more dynamic. Any ideas or is the calc method the best for this situation? The stepper in general seems to be tricky to style for with all of the nested elements.
```py
from nicegui import ui

with ui.card().classes("h-[90vh] w-[480px]"):
    with ui.stepper().classes("w-full h-full") as stepper:
        with ui.step("A").classes("w-full"):
            with ui.column().classes("w-full justify-between").style("height: calc(90vh - 128px)"):
                ui.input()
                with ui.stepper_navigation():
                    ui.button("Next", on_click=stepper.next)
        with ui.step("B").classes("w-full"):
            with ui.column().classes("w-full justify-between").style("height: calc(90vh - 128px)"):
                ui.input()
                with ui.stepper_navigation():
                    ui.button("Done")
ui.run()
```

**Comment by falkoschindler** [comment]
Good question, @natankeddem!

I experimented with some extra CSS definitions, but it is surprisingly complicated to get the desired result. The hierarchy is so deep and you have to touch almost every layer to make sure it fills the vertical space:
```py
ui.add_head_html('''
<style>
    .full-height-stepper,
    .full-height-stepper .q-stepper__content,
    .full-height-stepper .q-stepper__step-content,
    .full-height-stepper .q-stepper__step-inner {
        height: 100%;
        display: flex;
        flex-direction: column;
    }
</style>
''')

with ui.card().classes("h-[90vh] w-[480px]"):
    with ui.stepper().classes("w-full full-height-stepper") as stepper:
        with ui.step("A").classes("w-full"):
            ui.input()
            with ui.stepper_navigation().classes("mt-auto"):
                ui.button("Next", on_click=stepper.next)
        with ui.step("B").classes("w-full"):
            ui.input()
            with ui.stepper_navigation().classes("mt-auto"):
                ui.button("Done")
```
Using a flex layout allows for the stepper navigation to move to the bottom when setting a top margin of "auto". Of course, you could define the top margin via CSS as well:
```css
    .full-height-stepper .q-stepper__nav {
        margin-top: auto;
    }
```
So even if my approach might be more generic without magic numbers like 128px, yours is totally valid it should work as well - without the need for so much overhead.

I wonder if NiceGUI could add such CSS definitions by default so that `ui.stepper` and `ui.step` can be styled more intuitively. See https://github.com/zauberzeug/nicegui/issues/1614 where we're planning to use flex layout for tabs by default to get a `ui.column`-like behavior. Maybe we can do that for steppers as well.

**Comment by natankeddem** [comment]
I have resorted to utilizing similar magic numbers in the past on tab panel as well. I think using flexboxs in more places is a good thing. I know they are traditional more difficult to develop/debug but with NiceGUI's syntaxes they don't seem overwhelming.
